src/point_one/polaris/polaris_client.cc

```diff
@@ -0,0 +1,296 @@
+  std::unique_lock<std::recursive_mutex> lock(mutex_);
+  max_reconnect_attempts_ = max_reconnect_attempts;
+}
+
+/******************************************************************************/
+void PolarisClient::SetRTCMCallback(
+    std::function<void(const uint8_t* buffer, size_t size_bytes)> callback) {
+  std::unique_lock<std::recursive_mutex> lock(mutex_);
+  callback_ = callback;
+}
+
+/******************************************************************************/
+void PolarisClient::SendECEFPosition(double x_m, double y_m, double z_m) {
+  VLOG(1) << "Setting current ECEF position: [" << std::fixed
+          << std::setprecision(2) << x_m << ", " << y_m << ", " << z_m << "]";
+  std::unique_lock<std::recursive_mutex> lock(mutex_);
+  current_request_type_ = RequestType::ECEF;
+  ecef_position_m_[0] = x_m;
+  ecef_position_m_[1] = y_m;
+  ecef_position_m_[2] = z_m;
+  if (connected_) {
+    polaris_.SendECEFPosition(x_m, y_m, z_m);
+  }
+}
+
+/******************************************************************************/
+void PolarisClient::SendLLAPosition(double latitude_deg, double longitude_deg,
+                                    double altitude_m) {
+  VLOG(1) << "Setting current LLA position: [" << std::fixed
+          << std::setprecision(6) << latitude_deg << ", " << longitude_deg
+          << ", " << std::setprecision(2) << altitude_m << "]";
+  std::unique_lock<std::recursive_mutex> lock(mutex_);
+  current_request_type_ = RequestType::LLA;
+  lla_position_deg_[0] = latitude_deg;
+  lla_position_deg_[1] = longitude_deg;
+  lla_position_deg_[2] = altitude_m;
+  if (connected_) {
+    polaris_.SendLLAPosition(latitude_deg, longitude_deg, altitude_m);
+  }
+}
+
+/******************************************************************************/
+void PolarisClient::RequestBeacon(const std::string& beacon_id) {
+  VLOG(1) << "Requesting beacon '" << beacon_id << "'.";
+  std::unique_lock<std::recursive_mutex> lock(mutex_);
+  current_request_type_ = RequestType::BEACON;
+  beacon_id_ = beacon_id;
+  if (connected_) {
+    polaris_.RequestBeacon(beacon_id.c_str());
+  }
+}
+
+/******************************************************************************/
+void PolarisClient::Run(double timeout_sec) {
+  const int timeout_ms = std::lround(timeout_sec * 1e3);
+  running_ = true;
+  while (running_) {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+
+    // Retrieve an access token using the specified API key.
+    if (!auth_valid_) {
+      VLOG(1) << "Authenticating with Polaris service. [unique_id="
+              << unique_id_ << "]";
+      int ret = polaris_.Authenticate(api_key_, unique_id_);
+      if (ret == POLARIS_FORBIDDEN) {
+        LOG(ERROR) << "Authentication rejected. Is your API key valid?";
+        running_ = false;
+        break;
+      } else if (ret != POLARIS_SUCCESS) {
+        LOG(WARNING) << "Authentication failed. Retrying.";
```

Comment:
we should pipe out the value of ret here.

Reply:
We can't easily with the current implementation -- the whole point of this function is to block forever, so how would it output the return code? We could maybe have a callback that got called on any error?